### PR TITLE
Add rope dynamic linear scaling

### PR DESCRIPTION
**Pull request by hsiehjackson**
# What does this PR do ?

Add dynamic position interpolation for `seq length > learned length`
Learned length is `pre-trained max_position_embeddings * seq_len_interpolation_factor`

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py [hunk 1]
```diff
@@ -25,7 +25,9 @@ class RotaryEmbedding(nn.Module):
     Implements Rotary Position Embedding from https://arxiv.org/abs/2104.09864.
     """
 
-    def __init__(self, dim: int, seq_len_interpolation_factor: int = None):
+    def __init__(
+        self, dim: int, seq_len_interpolation_factor: int = None, pretrained_max_position_embeddings: int = None
+    ):
```
Comment (yzhang123) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py:30]: could you add doc strings?

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py [hunk 2]
```diff
@@ -34,16 +36,24 @@ def __init__(self, dim: int, seq_len_interpolation_factor: int = None):
...
 
     def forward(self, max_seq_len, offset=0):
         seq = torch.arange(max_seq_len, device=self.inv_freq.device) + offset
-        if self.seq_len_interpolation_factor is not None:
-            seq = seq.type_as(self.inv_freq)
-            seq *= 1 / self.seq_len_interpolation_factor
-        freqs = einsum('i , j -> i j', seq.type_as(self.inv_freq), self.inv_freq)
+        seq = seq.type_as(self.inv_freq)
+
+        if self.pretrained_max_position_embeddings is not None:
```
Comment (yzhang123) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py:48]: when is this ever none?

**Review by yzhang123 — CHANGES_REQUESTED** [review]
small comments, rest looks good. thx!

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py (hunk 1 above)
Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py:30]: Added.

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py (hunk 2 above)
Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py]: In retrieval model, we use Rope without this parameter. https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/modules/common/megatron/retrieval_transformer.py#L160

**Review by yzhang123 — APPROVED** [review]

**Comment by yzhang123** [comment]
will this PR be merged into sft-scrolls too?

**Pull request merged**